File: Python/0700. Search in a Binary Search Tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

#法一: recursion
class Solution(object):
    def searchBST(self, root, val):
        #搜索到目標節點就要立即return(若不加就是traverse整棵樹)

        if not root or root.val == val:
            return root

        if root.val > val:
            return self.searchBST(root.left, val)

        if root.val < val:
            return self.searchBST(root.right, val)

#法二: iteration
class Solution(object):
    def searchBST(self, root, val):
        while root:
            if val < root.val:
                root = root.left

            elif val > root.val:
                root = root.right

            else: return root

        return None

File: Java/7. Binary Tree/0700. Search in a Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// 一般遞迴: binary tree
class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        if (root == null || root.val == val) {
            return root;
        }
        TreeNode left = searchBST(root.left, val);
        if (left != null) {
            return left;
        }
        return searchBST(root.right, val);
    }
}

class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        if (root == null || root.val == val) {
            return root;
        }
        // 如果目標值小於當前節點的值，則往左子樹搜尋
        if (val < root.val) {
            return searchBST(root.left, val);
        } else { // 否則往右子樹搜尋
            return searchBST(root.right, val);
        }
    }
}